api/client/src/main/java/org/projectnessie/client/auth/oauth2/AbstractFlow.java

```diff
@@ -27,15 +27,31 @@ abstract class AbstractFlow implements Flow {
     this.config = config;
   }
 
-  <T> T invokeEndpoint(URI endpoint, Object request, Class<? extends T> responseClass) {
+  <REQ extends TokensRequestBase, RESP extends TokensResponseBase> RESP invokeTokenEndpoint(
+      TokensRequestBase.Builder<REQ> request, Class<? extends RESP> responseClass) {
+    request.scope(config.getScope().orElse(null));
+    if (request instanceof PublicTokensRequestBase.Builder
+        && !config.getClientSecret().isPresent()) {
```

Comment:
Why does it only include the client-ID, if the secret is not present? The comment in AuthorizationCodeFlow says it must always be present.

Reply:
The general recommendation for all grants is:

* For confidential clients:
  -  Authenticate the client with a Basic authentication header
  - do NOT include `client_id` nor `client_secret` in POST request body
    - While it is also possible to authenticate the client using `client_id` nor `client_secret` in POST request body, this method is considered less secure. The `OAuth2Client` does not use it.
 * For public clients:
   - Do not include Basic authentication header (since there is no secret)
   - But include `client_id` in POST request body (but no `client_secret` of course), in order to identify the client.

In `AuthorizationCodeFlow`, when constructing the authorization URI, the OAuth2 spec mandates that the client ID be included. My comment there is however a bit superfluous, I will remove it.

Reply:
Mind adding this to the javadoc? I'll likely forget soon ;)